### PR TITLE
BUG: Eval scopes ignoring empty dictionaries (#47084)

### LINKED ISSUE #47084: BUG: eval and query ignore empty local_dict and global_dict
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
df = pd.DataFrame([[1,2], [1,2]], columns=["a", "b"])
df.eval("c = @df.a + @df.b", local_dict={}, global_dict={})


```
   a  b  c
0  1  2  3
1  1  2  3
```
```

### Issue Description

If someone was to pass empty dictionaries into `local_dict` and `global_dict` kwargs of `pd.eval`, `df.eval` or `df.query`, it would work as if `None` was passed instead (i.e. actual scope would be populated with current frame locals and globals). This is an unexpected behavior and a security concern.

The reason behind this is an incorrect check of argument `None` equality [there](https://github.com/pandas-dev/pandas/blob/1be9d3868f4b9cfd083839cf39eecfd881f387ad/pandas/core/computation/scope.py#L136) and [there](https://github.com/pandas-dev/pandas/blob/1be9d3868f4b9cfd083839cf39eecfd881f387ad/pandas/core/computation/scope.py#L140). If user passes `global_dict={}`, the following happens: `global_dict or frame.f_globals => {} or frame.f_globals => False or frame.f_globals => frame.f_globals` and the same with `local_dict` later on.

### Expected Behavior

```python
import pandas as pd
df = pd.DataFrame([[1,2], [1,2]], columns=["a", "b"])
df.eval("c = @df.a + @df.b", local_dict={}, global_dict={})
```
```python-traceback
UndefinedVariableError: Undefined variable 'df'
```
Or a similar error

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 1be9d3868f4b9cfd083839cf39eecfd881f387ad
python           : 3.9.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.17.5-76051705-generic
Version          : #202204271406~1651504840~21.10~63e51bd SMP PREEMPT Mon May 2 15:
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.5.0.dev0+798.g1be9d3868f
numpy            : 1.21.3
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 20.3.4
setuptools       : 52.0.0
Cython           : 0.29.30
pytest           : 6.2.5
hypothesis       : 6.46.7
sphinx           : 4.5.0
blosc            : 1.10.6
feather          : None
xlsxwriter       : 3.0.3
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.1.1
IPython          : 7.33.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.4
brotli           : None
fastparquet      : 0.7.2
fsspec           : 2022.5.0
gcsfs            : 2022.5.0
matplotlib       : 3.4.3
numba            : 0.53.1
numexpr          : 2.8.1
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : 7.0.0
pyreadstat       : 1.1.6
pyxlsb           : None
s3fs             : 0.6.0
scipy            : 1.7.3
snappy           : 
sqlalchemy       : 1.4.36
tables           : 3.7.0
tabulate         : 0.8.9
xarray           : 2022.3.0
xlrd             : 2.0.1
xlwt             : 1.3.0
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by Alex-Blade**
closes  [#47084](https://github.com/pandas-dev/pandas/issues/47084)

- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jreback** [comment]
looks ok, need to do precommit here (line length). can you add a whatsnew note for this.

**Comment by Alex-Blade** [comment]
@github-actions pre-commit

**Comment by Alex-Blade** [comment]
@jreback, to be honest, I don't really get how should I name the entry in the `_whatsnew_150.bug_fixes` section. Some of them are named according to classnames (`Categorical` - Bug in `Categorical.view`). Others imply an action during which users may encounter the bug (`Indexing` - Bug in `loc.__getitem__`). Is there a list of predefined sections perhaps?

**Comment by Alex-Blade** [comment]
UPD. Made an entry in Other section, by analogy with similar bug-fixes from previous versions

File: doc/source/whatsnew/v1.5.0.rst [hunk 1]
```diff
@@ -824,6 +824,7 @@ Metadata
 
 Other
 ^^^^^
+- Bug in :meth:`pd.eval`, :meth:`DataFrame.eval` and :meth:`DataFrame.query` where passing empty ``local_dict`` or ``global_dict`` was treated as passing ``None`` (:issue:`47084`)
```
[the remark this answers is not in the document]
Comment (Alex-Blade) [reply doc/source/whatsnew/v1.5.0.rst:827]: Under ".. ***DO NOT USE THIS SECTION***"? Feels wrong

[the remark this answers is not in the document]
Comment (Alex-Blade) [reply doc/source/whatsnew/v1.5.0.rst]: Got it, Done

File: doc/source/whatsnew/v1.5.0.rst [hunk 2]
```diff
@@ -827,7 +827,7 @@ Other
 
 .. ***DO NOT USE THIS SECTION***
 
--
+- Bug in :meth:`pd.eval`, :meth:`DataFrame.eval` and :meth:`DataFrame.query` where passing empty ``local_dict`` or ``global_dict`` was treated as passing ``None`` (:issue:`47084`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:830]: move to conversion

File: pandas/tests/computation/test_eval.py [hunk 3]
```diff
@@ -1659,6 +1660,18 @@ def test_no_new_globals(self, engine, parser):
         gbls2 = globals().copy()
         assert gbls == gbls2
 
+    def test_empty_locals(self, engine, parser):
+        x = 1
```
Comment (jreback) [inline pandas/tests/computation/test_eval.py:1664]: can you add the issue number as a comment in these 2 new tests

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by Alex-Blade** [comment]
@jreback, mind looking at it again?

**Review by jreback — COMMENTED** [review]
could do for 1.4.x

File: doc/source/whatsnew/v1.5.0.rst [hunk 4]
```diff
@@ -675,6 +675,7 @@ Conversion
 - Bug when comparing string and datetime64ns objects causing ``OverflowError`` exception. (:issue:`45506`)
 - Bug in metaclass of generic abstract dtypes causing :meth:`DataFrame.apply` and :meth:`Series.apply` to raise for the built-in function ``type`` (:issue:`46684`)
 - Bug in :meth:`DataFrame.to_dict` for ``orient="list"`` or ``orient="index"`` was not returning native types (:issue:`46751`)
+- Bug in :meth:`pd.eval`, :meth:`DataFrame.eval` and :meth:`DataFrame.query` where passing empty ``local_dict`` or ``global_dict`` was treated as passing ``None`` (:issue:`47084`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.5.0.rst:678]: @Alex-Blade can you move to bug fixes section of doc/source/whatsnew/v1.4.3.rst

Comment (Alex-Blade) [reply doc/source/whatsnew/v1.5.0.rst]: Done

**Comment by jreback** [comment]
see here: https://github.com/pandas-dev/pandas/runs/6660273116?check_suite_focus=true

looks like the precommit is failing

File: pandas/core/computation/pytables.py [hunk 5]
```diff
@@ -563,7 +563,7 @@ def __init__(
         self._visitor = None
 
         # capture the environment if needed
-        local_dict: DeepChainMap[Any, Any] = DeepChainMap()
+        local_dict: Optional[DeepChainMap[Any, Any]] = None
```
Comment (jreback) [inline pandas/core/computation/pytables.py:566]: did this change any tests?

**Comment by Alex-Blade** [comment]
@github-actions pre-commit

File: pandas/core/computation/pytables.py (hunk 5 above)
Comment (Alex-Blade) [reply pandas/core/computation/pytables.py]: Oh, I was waiting for the test and didn't see your message. Yes, turned out this part of code relied on `bool(DeepChainMap())` being equal to `False`. This fixed all the issues, waiting for pre-commit now

**Comment by Alex-Blade** [comment]
@github-actions pre-commit

**Comment by Alex-Blade** [comment]
Seems like Numpy Dev died by an accident on setup-miniconda stage

**Comment by Alex-Blade** [comment]
Those test are painful... Two times they died because of numpy, and now a timeout...

**Comment by Alex-Blade** [comment]
@simonjayhawkins, it has passed the tests a few times in a row, I think it can be considered as ready

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Alex-Blade very nice!

**Comment by jreback** [comment]
@meeseeksdev backport 1.4.x

**Pull request merged**